git/libgit2/repository_client.go

```diff
@@ -0,0 +1,537 @@
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			rel, err := filepath.Rel(l.repository.Workdir(), path)
+			if err != nil {
+				return err
+			}
+			if info.IsDir() || strings.HasPrefix(rel, ".git") || rel == "." {
+				return nil
+			}
+
+			if err := index.AddByPath(rel); err != nil {
+				return err
+			}
+			return nil
+		}); err != nil {
+		return "", err
+	}
+
+	if err := index.Write(); err != nil {
+		return "", err
+	}
+
+	treeID, err := index.WriteTree()
+	if err != nil {
+		return "", err
+	}
+
+	tree, err := l.repository.LookupTree(treeID)
+	if err != nil {
+		return "", err
+	}
+	defer tree.Free()
+
+	sig := &git2go.Signature{
+		Name:  info.Author.Name,
+		Email: info.Author.Email,
+		When:  time.Now(),
+	}
+	commitID, err := l.repository.CreateCommit("HEAD", sig, sig, info.Message, tree, parentC...)
+	if err != nil {
+		return "", err
+	}
+
+	// return unsigned commit if pgp entity is not provided
+	if signer == nil {
+		return commitID.String(), nil
+	}
+
+	commit, err := l.repository.LookupCommit(commitID)
+	if err != nil {
+		return "", err
+	}
+	defer commit.Free()
+
+	signedCommitID, err := commit.WithSignatureUsing(func(commitContent string) (string, string, error) {
+		cipherText := new(bytes.Buffer)
+		err := openpgp.ArmoredDetachSignText(cipherText, signer, strings.NewReader(commitContent), &packet.Config{})
+		if err != nil {
+			return "", "", errors.New("error signing payload")
+		}
+
+		return cipherText.String(), "", nil
+	})
+	if err != nil {
+		return "", err
+	}
+	signedCommit, err := l.repository.LookupCommit(signedCommitID)
+	if err != nil {
+		return "", err
+	}
+	defer signedCommit.Free()
+
+	newHead, err := l.repository.Head()
+	if err != nil {
+		return "", err
+	}
+	defer newHead.Free()
+
+	signedRef, err := l.repository.References.Create(
+		newHead.Name(),
+		signedCommit.Id(),
+		true,
+		"repoint to signed commit",
+	)
+	if err != nil {
+		return "", err
+	}
+	signedRef.Free()
+
+	return signedCommitID.String(), nil
+}
+
+func (l *Client) Push(ctx context.Context) error {
+	if l.repository == nil {
+		return git.ErrNoGitRepository
+	}
+	callbacks := RemoteCallbacks()
+
+	head, err := l.repository.Head()
+	if err != nil {
+		return err
+	}
+	defer head.Free()
+	branch, err := head.Branch().Name()
+	if err != nil {
+		return err
+	}
+	// calling repo.Push will succeed even if a reference update is
+	// rejected; to detect this case, this callback is supplied.
+	var callbackErr error
+	callbacks.PushUpdateReferenceCallback = func(refname, status string) error {
+		if status != "" {
+			callbackErr = fmt.Errorf("ref %s rejected: %s", refname, status)
+		}
+		return nil
+	}
+	err = l.remote.Push([]string{fmt.Sprintf("refs/heads/%s:refs/heads/%[1]s", branch)}, &git2go.PushOptions{
+		RemoteCallbacks: callbacks,
+		ProxyOptions:    git2go.ProxyOptions{Type: git2go.ProxyTypeAuto},
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "early EOF") {
```

Comment:
We also need a test for this similar to the one in https://github.com/fluxcd/image-automation-controller/pull/410 maybe adapted to the tests in this package.